File: merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((l1, l2) -> Integer.compare(l1.val, l2.val));
        
        for (ListNode l : lists) {
            if (l != null) {
                pq.add(l);
            }
        }
        
        ListNode dummyHead = new ListNode();
        ListNode prev = dummyHead;
        
        while (!pq.isEmpty()) {
            ListNode l = pq.poll();
            prev.next = new ListNode(l.val);
            prev = prev.next;
            
            if (l.next != null) {
                pq.add(l.next);
            }
        }
        
        return dummyHead.next;
    }
}

File: merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        h = []
        i = 0
        for head in lists:
            if head is None:
                continue
            heapq.heappush(h, (head.val, i, head))
            i += 1
        
        dummy_head = ListNode()
        prev = dummy_head
        while h:
            _, _, node = heapq.heappop(h)
            prev.next = node
            prev = node
            
            if node.next is not None:
                heapq.heappush(h, (node.next.val, i, node.next))
                i += 1
        
        return dummy_head.next
            